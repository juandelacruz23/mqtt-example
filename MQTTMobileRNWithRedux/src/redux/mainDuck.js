/**
 * What is a duck?
 * https://github.com/erikras/ducks-modular-redux
 */

import { connectionStatuses, subscriptionStatuses } from "../statuses";

const CHANGE_AND_PUSH = "CHANGE_AND_PUSH";
const CHANGE_VALUE = "CHANGE_VALUE";
const PUSH_TEXT = "PUSH_TEXT";

export const changeAndPush = (newValue, textToPush) => ({
  type: CHANGE_AND_PUSH,
  payload: {
    value: newValue,
    text: textToPush,
  },
});

export const changeValue = newValue => ({
  type: CHANGE_VALUE,
  payload: newValue,
});

export const pushText = text => ({
  type: PUSH_TEXT,
  payload: text,
});

const initialState = {
  text: [],
  connectionStatus: connectionStatuses.DISCONNECTED,
  subscriptionStatus: subscriptionStatuses.UNSUBSCRIBED,
  host: "",
  port: "",
  topic: "",
  showMessageDialog: false,
  message: "",
  connecting: false,
};

export function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case CHANGE_AND_PUSH:
      return {
        ...state,
        ...action.payload.value,
        text: [...state.text, action.payload.text],
      };
    case CHANGE_VALUE:
      return { ...state, ...action.payload };
    case PUSH_TEXT:
      return { ...state, text: [...state.text, action.payload] };
    default:
      return state;
  }
}
