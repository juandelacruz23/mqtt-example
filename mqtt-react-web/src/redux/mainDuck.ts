import { createAction } from "@reduxjs/toolkit";
import HistoryItem from "../types/HistoryItem";
import MQTTOptions from "../types/MQTTOptions";
import Subscription from "../types/Subscription";
import ConsoleEvent from "../types/ConsoleEvents";

const CHANGE_VALUE = "CHANGE_VALUE";
export const CONNECT = "CONNECT";
export const DISCONNECT = "DISCONNECT";
export const CONSOLE_EVENT = "CONSOLE_EVENT";
export const MESSAGE_RECEIVED = "MESSAGE_RECEIVED";
export const SUBSCRIBE = "SUBSCRIBE";
export const UNSUBSCRIBE = "UNSUBSCRIBE";

export interface BaseAction {
  readonly type: string;
  readonly payload: object;
}

export interface StringAction {
  readonly type: string;
  readonly payload: string;
}

export interface PayloadlessAction {
  readonly type: string;
}

export const INITIAL_STATE = {
  host: "",
  isConnected: false,
  port: 0,
  clientId: ``,
  path: "",
  events: [
    {
      value: "INFO - Starting Eclipse Paho JavaScript Utility.",
      timestamp: Date.now(),
    },
  ],
  messages: [],
  topic: "message",
};

// Little hack to override the "messages" type
export type AppState = Omit<typeof INITIAL_STATE, "messages"> & {
  messages: HistoryItem[];
};

export type AppAction = BaseAction | PayloadlessAction | StringAction;

export const changeValue = createAction<Partial<AppState>>(CHANGE_VALUE);

export const connectClient = createAction<MQTTOptions>(CONNECT);

export const consoleEvent = createAction<ConsoleEvent>(CONSOLE_EVENT);

export const disconnectClient = createAction(DISCONNECT);

export const messageReceived = createAction<HistoryItem>(MESSAGE_RECEIVED);

export const subscribe = createAction<Subscription>(SUBSCRIBE);

export const unsubscribe = createAction<string>(UNSUBSCRIBE);

export function reducer(
  state: AppState = INITIAL_STATE,
  action: BaseAction,
): AppState {
  switch (action.type) {
    case changeValue.type:
    case connectClient.type:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
}

export default {
  connectClient,
  changeValue,
  disconnectClient,
};
