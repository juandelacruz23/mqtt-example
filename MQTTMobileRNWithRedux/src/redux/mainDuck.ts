/**
 * What is a duck?
 * https://github.com/erikras/ducks-modular-redux
 */

import { Action } from "redux";
import ConnectionStatus from "../ConnectionStatus";
import SubscriptionStatus from "../SubscriptionStatus";

const CHANGE_AND_PUSH = "CHANGE_AND_PUSH";
const CHANGE_VALUE = "CHANGE_VALUE";
const PUSH_TEXT = "PUSH_TEXT";

export interface AppAction extends Action<string> {
  type: string,
  payload?: any,
}

export const changeAndPush = (newValue: Object, textToPush: string) : AppAction => ({
  type: CHANGE_AND_PUSH,
  payload: {
    value: newValue,
    text: textToPush,
  },
});

export const changeValue = (newValue: Object) : AppAction => ({
  type: CHANGE_VALUE,
  payload: newValue,
});

export const pushText = (text: string): AppAction => ({
  type: PUSH_TEXT,
  payload: text,
});

export interface AppState {
  connectionStatus: ConnectionStatus,
  host: string,
  loading: boolean,
  message: string,
  port: string,
  showMessageDialog: boolean,
  subscriptionStatus: SubscriptionStatus,
  text: string[],
  topic: string,
}

const initialState: AppState = {
  text: [],
  connectionStatus: ConnectionStatus.DISCONNECTED,
  subscriptionStatus: SubscriptionStatus.UNSUBSCRIBED,
  host: "",
  port: "",
  topic: "",
  showMessageDialog: false,
  message: "",
  loading: false,
};

export function reducer(state = initialState, action: AppAction = { type: 'DEFAULT'} ): AppState {
  switch (action.type) {
    case CHANGE_AND_PUSH:
      return {
        ...state,
        ...action.payload.value,
        text: [...state.text, action.payload.text],
      };
    case CHANGE_VALUE:
      return { ...state, ...action.payload };
    case PUSH_TEXT:
      return { ...state, text: [...state.text, action.payload] };
    default:
      return state;
  }
}
