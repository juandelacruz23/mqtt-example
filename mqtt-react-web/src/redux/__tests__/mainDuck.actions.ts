import { changeValue } from "../mainDuck";
describe("change value action", () => {
  it("should return CHANGE_VALUE with payload", () => {
    const fakePath = "fake path";
    const firstAction = changeValue({ path: fakePath });
    expect(firstAction).toEqual({
      type: "CHANGE_VALUE",
      payload: {
        path: fakePath,
      },
    });
    expect(firstAction).toMatchSnapshot();

    const fakeTopic = "fakeTopic";
    const secondAction = changeValue({ topic: fakeTopic });
    expect(secondAction).toEqual({
      type: "CHANGE_VALUE",
      payload: {
        topic: fakeTopic,
      },
    });
    expect(secondAction).toMatchSnapshot();
  });
});
