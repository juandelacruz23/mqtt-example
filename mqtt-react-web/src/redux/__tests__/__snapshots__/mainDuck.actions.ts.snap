// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change value action should return CHANGE_VALUE with payload 1`] = `
Object {
  "payload": Object {
    "path": "fake path",
  },
  "type": "CHANGE_VALUE",
}
`;

exports[`change value action should return CHANGE_VALUE with payload 2`] = `
Object {
  "payload": Object {
    "topic": "fakeTopic",
  },
  "type": "CHANGE_VALUE",
}
`;
